pkg/httpclient/http.go

```diff
@@ -0,0 +1,130 @@
+// limitations under the License.
+
+package httpclient
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"runtime"
+	"time"
+
+	"github.com/gregjones/httpcache"
+	"go.thethings.network/lorawan-stack/v3/pkg/config/tlsconfig"
+	"go.thethings.network/lorawan-stack/v3/pkg/version"
+)
+
+// defaultHTTPClientTimeout is the default timeout for the HTTP client.
+const defaultHTTPClientTimeout = 10 * time.Second
+
+// TLSClientConfigurationProvider provides a *tls.Config to be used by TLS clients.
+type TLSClientConfigurationProvider interface {
+	GetTLSClientConfig(context.Context, ...tlsconfig.Option) (*tls.Config, error)
+}
+
+// Provider constructs *http.Clients.
+type Provider interface {
+	HTTPClient(context.Context, ...Option) (*http.Client, error)
+}
+
+// Option is an option for HTTP clients.
+type Option func(*httpClientOptions)
+
+type httpClientOptions struct {
+	cache            bool
+	tlsConfig        *tls.Config
+	tlsConfigOptions []tlsconfig.Option
+}
+
+// WithCache enables caching at transport level.
+func WithCache(b bool) Option {
+	return Option(func(o *httpClientOptions) {
+		o.cache = true
+	})
+}
+
+// WithTLSConfig configures the TLS configuration to be used by the transport.
+func WithTLSConfig(c *tls.Config) Option {
+	return Option(func(o *httpClientOptions) {
+		o.tlsConfig = c
+	})
+}
+
+// WithTLSConfigOptions configures the TLS configuration options provided to the TLS configuration provider
+// by the transport.
+func WithTLSConfigOptions(opts ...tlsconfig.Option) Option {
+	return Option(func(o *httpClientOptions) {
+		o.tlsConfigOptions = opts
+	})
+}
+
+type provider struct {
+	tlsConfigProvider TLSClientConfigurationProvider
+}
+
+// HTTPClient returns a new *http.Client with a default timeout and a configured transport.
+func (p *provider) HTTPClient(ctx context.Context, opts ...Option) (*http.Client, error) {
+	options := &httpClientOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if options.tlsConfig == nil {
+		tlsConfig, err := p.tlsConfigProvider.GetTLSClientConfig(ctx, options.tlsConfigOptions...)
+		if err != nil {
+			return nil, err
+		}
+		options.tlsConfig = tlsConfig
+	}
+
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.TLSClientConfig = options.tlsConfig
+
+	rt := http.RoundTripper(transport)
+	if options.cache {
+		rt = &httpcache.Transport{
+			Transport:           rt,
+			Cache:               httpcache.NewMemoryCache(),
+			MarkCachedResponses: true,
+		}
+	}
+	rt = &roundTripperWithUserAgent{
+		RoundTripper: rt,
+		UserAgent:    fmt.Sprintf("TheThingsStack/%s (%s/%s)", version.TTN, runtime.GOOS, runtime.GOARCH),
+	}
+
+	return &http.Client{
+		Timeout:   defaultHTTPClientTimeout,
+		Transport: rt,
+	}, nil
+}
+
+// NewProvider constructs a Provider on top of the provided TLS configuration provider.
+func NewProvider(tlsConfigProvider TLSClientConfigurationProvider) Provider {
+	return &provider{tlsConfigProvider: tlsConfigProvider}
+}
+
```

Comment:
Wouldn't it be possible to have the tls configuration as an `Option` for the `HTTPClient` method? That way you wouldn't need the Provider interface, but rather the `tls` package could have a public function that would create the `Option` func for the creation of the `httpclient` instance.

It's a small change but you would remove the `tlsconfig` definitions/import from the package.

Something like:
```go
// in the `tlsconfig` pkg 
func (p ConfigurationProvider) GetTLSClientConfig(ctx context.Context, opts ...Option) (httpclient.Option, error)

// in component/http.go
func (c *Component) HTTPClient(ctx context.Context, opts ...httpclient.Option) (*http.Client, error) {
        tlsOpt, err := GetTLSClientConfig(ctx, opts...)
	return httpclient.HTTPClient(ctx, tlsOpt)
}
```

Reply:
The currently implemented variant allows callers to request different `tlsconfig.Option` directly on `HTTPClient` using the `WithTLSConfigOptions` option. If we explicitly construct the `*tls.Config` in the component, you cannot actually pass these without reinventing `WithTLSConfigOptions`, which makes `httpclient` depened on `tlsconfig`.

Reply:
I suppose my question was an extension of the idea of making the options struct exported, that way these functions wouldn't need to exist. Since the initial idea is quite troublesome due to the capability of being used outside the pkg, I believe this comment doesn't make sense anymore.